src/data/polynomial/laurent.lean

```diff
@@ -0,0 +1,314 @@
+
+@[simp] lemma polynomial_one_eq_one : (polynomial.to_laurent : R[X] → R[T;T⁻¹]) 1 = 1 :=
+C_eq_C (1 : R)
+
+@[simp]
+lemma polynomial_C_mul_eq (r : R) (f : R[X]):
+  (polynomial.C r * f).to_laurent = C r * f.to_laurent :=
+by simp only [_root_.map_mul, C_eq_C]
+
+@[simp]
+lemma X_pow_eq_T (n : ℕ) :
+  (X ^ n : R[X]).to_laurent = T n :=
+by simp only [map_pow, X_eq_T, T_pow, mul_one]
+
+@[simp]
+lemma C_mul_X_pow (n : ℕ) (r : R) :
+  (polynomial.C r * X ^ n).to_laurent = C r * T n :=
+by simp only [_root_.map_mul, C_eq_C, X_pow_eq_T]
+
+lemma is_unit_T (n : ℤ) : is_unit (T n : R[T;T⁻¹]) :=
+by refine ⟨⟨T n, T (- n), _, _⟩, _⟩; simp [← T_add]
+
+lemma is_regular_T (n : ℤ) : is_regular (T n : R[T;T⁻¹]) :=
+⟨is_left_regular_of_mul_eq_one (by simp [← T_add] : T (-n) * T n = 1),
+  is_right_regular_of_mul_eq_one (by simp [← T_add] : T n * T (-n) = 1)⟩
+
+@[elab_as_eliminator] protected lemma induction_on {M : R[T;T⁻¹] → Prop} (p : R[T;T⁻¹])
+  (h_C         : ∀ a, M (C a))
+  (h_add       : ∀ {p q}, M p → M q → M (p + q))
+  (h_C_mul_T   : ∀ (n : ℕ) (a : R), M (C a * T n) → M (C a * T (n + 1)))
+  (h_C_mul_T_Z : ∀ (n : ℕ) (a : R), M (C a * T (- n)) → M (C a * T (- n - 1))) :
+  M p :=
+begin
+  have A : ∀ {n : ℤ} {a : R}, M (C a * T n),
+  { assume n a,
+    apply n.induction_on,
+    { simpa only [T_zero, mul_one] using h_C a },
+    { exact λ m, h_C_mul_T m a },
+    { exact λ m, h_C_mul_T_Z m a } },
+  have B : ∀ (s : finset ℤ), M (s.sum (λ (n : ℤ), C (p.to_fun n) * T n)),
+  { apply finset.induction,
+    { convert h_C 0, simp only [finset.sum_empty, _root_.map_zero] },
+    { assume n s ns ih, rw finset.sum_insert ns,
+    { exact h_add A ih } } },
+  convert B p.support,
+  ext a,
+  simp_rw [← single_eq_C_mul_T, finset.sum_apply', single_apply, finset.sum_ite_eq'],
+  split_ifs with h h,
+  { refl },
+  { exact finsupp.not_mem_support_iff.mp h }
+end
+
+/--  To prove something about Laurent polynomials, it suffices to show that
+* the condition is closed under taking sums, and
+* it holds for monomials.
+-/
+@[elab_as_eliminator] protected lemma induction_on' {M : R[T;T⁻¹] → Prop} (p : R[T;T⁻¹])
+  (h_add     : ∀p q, M p → M q → M (p + q))
+  (h_C_mul_T : ∀(n : ℤ) (a : R), M (C a * T n)) :
+  M p :=
+begin
+  refine p.induction_on (λ a, _) h_add _ _;
+  try { exact λ n f Mf, h_C_mul_T _ f },
+  convert h_C_mul_T 0 a,
+  exact (mul_one _).symm,
+end
+
+lemma commute_T (n : ℤ) (f : R[T;T⁻¹]) : commute (T n) f :=
+f.induction_on' (λ p q Tp Tq, commute.add_right Tp Tq) $ λ m a,
+show T n * _ = _, by
+{ rw [T, T, ← single_eq_C, single_mul_single, single_mul_single, single_mul_single],
+  simp [add_comm] }
+
+@[simp]
+lemma T_mul (n : ℤ) (f : R[T;T⁻¹]) : T n * f = f * T n :=
+(commute_T n f).eq
+
+/--  `trunc : R[T;T⁻¹] →+ R[X]` maps a Laurent polynomial `f` to the polynomial whose terms of
+strictly positive degree coincide with the ones of `f`.  The coefficients of the terms of
+non-positive degree of `f` are summed up and used as constant term.
```

Comment:
This feels like an undesirable behavior!

Reply:
I agree, I would have wanted the map to simply be the projection onto the terms of non-negative degree, but I was not able to find a succint way of formalizing this!

Reply:
Not at lean right now, but if I had to guess just from looking at the code can you use comap_domain rather than map_domain? Assuming it exists...

Reply:
Alex, I think that you are right, but while there is a `map_domain.add_monoid_hom` already available, the corresponding bundling for `comap_domain` is not in mathlib...